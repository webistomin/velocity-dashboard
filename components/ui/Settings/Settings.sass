+b('settings')
  +e('container')
    max-width: rem(1140px)

  +e('title')
    margin-bottom: rem(14px)

  +e('grid')
    display: grid
    grid-gap: rem(40px)

    +media($screen-sm)
      grid-gap: rem(50px)

    +media($screen-lg)
      grid-gap: rem(60px)

  +e('legend')
    margin-bottom: rem(24px)

  +e('inputs')
    display: grid
    grid-gap: rem(14px)
    margin-bottom: rem(24px)

    +media($screen-sm)
      grid-gap: rem(30px)
      margin-bottom: rem(30px)

  +e('input-row')
    display: grid
    grid-gap: rem(14px)

    +media($screen-sm)
      grid-template-columns: repeat(3, minmax(150px, 272px))

  +e('submit')
    +media($screen-lg)
      width: rem(272px)

  +e('input')
    +media($screen-sm)
      max-width: 272px

  +e('toggles')
    display: grid
    grid-gap: rem(20px)

    +media($screen-sm)
      grid-template-columns: repeat(2, minmax(150px, 346px))
      grid-gap: rem(40px)

    +media($screen-lg)
      grid-gap: rem(20px) rem(110px)

  +e('toggle-group')
    display: grid
    grid-gap: rem(20px)
    max-width: rem(346px)

    &:focus-within
      outline: rem(1px) solid rgb(var(--color-primary))
      outline-offset: 1rem

  +e('theme-group')
    display: grid
    grid-gap: rem(25px)
    grid-template-columns: repeat(2, minmax(100px, 140px))

    +media($screen-sm)
      grid-template-columns: repeat(5, minmax(100px, 140px))

    +media($screen-lg)
      grid-gap: rem(35px)

