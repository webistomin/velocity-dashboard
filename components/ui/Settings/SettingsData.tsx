import { VueComponent } from 'types/vue-components';
import { Component } from 'nuxt-property-decorator';
import { VNode } from 'vue';
import { email, sameAs } from 'vuelidate/lib/validators';

import BaseFormGroup from 'components/base/BaseFormGroup';
import BaseButton from 'components/base/BaseButton';

@Component({
  name: 'SettingsData',
  validations: {
    settingsForm: {
      email: {
        email,
      },
      firstName: {},
      lastName: {},
      dob: {},
      currentPassword: {},
      newPassword: {},
      confirmPassword: {
        sameAsPassword: sameAs('newPassword'),
      },
    },
  },
})
export default class SettingsData extends VueComponent {
  settingsForm = {
    email: '',
    firstName: '',
    lastName: '',
    dob: '',
    currentPassword: '123',
    newPassword: '',
    confirmPassword: '',
  };

  render(): VNode {
    return (
      <fieldset class='settings__fieldset fieldset'>
        <legend class='settings__legend legend'>
          Use this page to update your contact information and change your password.
        </legend>
        <div class='settings__inputs'>
          <div class='settings__input-row'>
            <BaseFormGroup
              class='settings__input'
              type='email'
              label='Email'
              id='settings-email'
              placeholder='Enter your email'
              validator={this.$v.settingsForm.email}
              onInput={($event: string) => (this.settingsForm.email = $event)}
              value={this.settingsForm.email}
              onBlur={this.$v.settingsForm.email?.$touch}
            />
          </div>
          <div class='settings__input-row'>
            <BaseFormGroup
              class='settings__input'
              type='text'
              label='First name'
              id='settings-first-name'
              placeholder='Enter your first name'
              validator={this.$v.settingsForm.firstName}
              onInput={($event: string) => (this.settingsForm.firstName = $event)}
              value={this.settingsForm.firstName}
              onBlur={this.$v.settingsForm.firstName?.$touch}
            />
            <BaseFormGroup
              class='settings__input'
              type='text'
              label='Last name'
              id='settings-last-name'
              placeholder='Enter your last name'
              validator={this.$v.settingsForm.lastName}
              onInput={($event: string) => (this.settingsForm.lastName = $event)}
              value={this.settingsForm.lastName}
              onBlur={this.$v.settingsForm.lastName?.$touch}
            />
            <BaseFormGroup
              class='settings__input'
              type='text'
              label='Birth date'
              id='settings-birth-date'
              placeholder='DD/MM/YYYY'
              validator={this.$v.settingsForm.dob}
              onInput={($event: string) => (this.settingsForm.dob = $event)}
              value={this.settingsForm.dob}
              onBlur={this.$v.settingsForm.dob?.$touch}
            />
          </div>
          <div class='settings__input-row'>
            <BaseFormGroup
              class='settings__input'
              type='password'
              label='Current password'
              id='settings-current-password'
              placeholder=''
              validator={this.$v.settingsForm.currentPassword}
              onInput={($event: string) => (this.settingsForm.currentPassword = $event)}
              value={this.settingsForm.currentPassword}
              onBlur={this.$v.settingsForm.currentPassword?.$touch}
            />
            <BaseFormGroup
              class='settings__input'
              type='text'
              label='New password'
              id='settings-new-password'
              placeholder='Enter new password'
              validator={this.$v.settingsForm.newPassword}
              onInput={($event: string) => (this.settingsForm.newPassword = $event)}
              value={this.settingsForm.newPassword}
              onBlur={this.$v.settingsForm.newPassword?.$touch}
            />
            <BaseFormGroup
              class='settings__input'
              type='text'
              label='Confirm'
              id='settings-confirm-password'
              placeholder='Repeat new password'
              validator={this.$v.settingsForm.confirmPassword}
              onInput={($event: string) => (this.settingsForm.confirmPassword = $event)}
              value={this.settingsForm.confirmPassword}
              onBlur={this.$v.settingsForm.confirmPassword?.$touch}
            />
          </div>
        </div>
        <BaseButton class='settings__submit' type='submit'>
          Save changes
        </BaseButton>
      </fieldset>
    );
  }
}
