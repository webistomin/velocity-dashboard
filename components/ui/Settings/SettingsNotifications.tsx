import { VueComponent } from 'types/vue-components';
import { Component, Emit, Prop } from 'nuxt-property-decorator';
import { VNode } from 'vue';

import BaseSwitch from 'components/base/BaseSwitch';
import { IUserSettings } from 'components/ui/Settings/Settings';

export interface ISettingsNotificationsProps {
  notifications: IUserSettings['notifications'];
  onSelectNotification?: (notifications: IUserSettings['notifications']) => void;
}

@Component({
  name: 'SettingsNotifications',
})
export default class SettingNotification extends VueComponent<ISettingsNotificationsProps> {
  @Prop()
  private readonly notifications!: IUserSettings['notifications'];

  public settingsForm: IUserSettings['notifications'] = this.notifications;

  @Emit('selectNotification')
  onInput(event: Event, field: keyof IUserSettings['notifications']): IUserSettings['notifications'] {
    const target = event.target as HTMLInputElement;
    this.settingsForm[field] = target.checked;
    return this.settingsForm;
  }

  public render(): VNode {
    return (
      <fieldset class='settings__fieldset fieldset'>
        <legend class='settings__legend legend'>Control your notification and auto-follow settings.</legend>
        <div class='settings__toggles'>
          <div class='settings__toggle-group'>
            <BaseSwitch
              label='Email Notification'
              text='Commits data and history'
              id='settings-email-notification'
              checked={this.settingsForm.isEmailNotificationsEnabled}
              onInput={(e: Event) => this.onInput(e, 'isEmailNotificationsEnabled')}
            />
            <BaseSwitch
              label='Push Notification'
              text='Commits data and history'
              id='settings-push-notification'
              checked={this.settingsForm.isPushNotificationsEnabled}
              onInput={(e: Event) => this.onInput(e, 'isPushNotificationsEnabled')}
            />
          </div>
          <div class='settings__toggle-group'>
            <BaseSwitch
              label='Monthly Reports'
              text='Commits data and history'
              id='settings-monthly-notification'
              checked={this.settingsForm.isMonthlyNotificationsEnabled}
              onInput={(e: Event) => this.onInput(e, 'isMonthlyNotificationsEnabled')}
            />
            <BaseSwitch
              label='Quarter Reports'
              text='Commits data and history'
              id='settings-quarter-notification'
              checked={this.settingsForm.isQuarterNotificationsEnabled}
              onInput={(e: Event) => this.onInput(e, 'isQuarterNotificationsEnabled')}
            />
          </div>
        </div>
      </fieldset>
    );
  }
}
