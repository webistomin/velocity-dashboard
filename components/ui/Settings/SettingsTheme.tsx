import { VueComponent } from 'types/vue-components';
import { Component, Emit, Prop } from 'nuxt-property-decorator';
import { VNode } from 'vue';

import BaseTheme from 'components/base/BaseTheme/BaseTheme';
import { IUserSettings } from 'components/ui/Settings/Settings';
import { SiteThemes } from 'common/types/theme/site-themes';

export interface ISettingsThemeProps {
  theme: IUserSettings['theme'];
  onThemeSelect?: (theme: IUserSettings['theme']) => void;
}

@Component({
  name: 'SettingsTheme',
})
export default class SettingsTheme extends VueComponent<ISettingsThemeProps> {
  @Prop()
  private readonly theme!: IUserSettings['theme'];

  public currentTheme: IUserSettings['theme'] = this.theme;

  @Emit('themeSelect')
  onInput(event: Event): IUserSettings['theme'] {
    const target = event.target as HTMLInputElement;
    this.currentTheme = target.value as SiteThemes;
    return this.currentTheme;
  }

  public render(): VNode {
    return (
      <fieldset class='settings__fieldset fieldset'>
        <legend class='settings__legend legend'>Select a color scheme for you Velocity app.</legend>
        <div class='settings__theme-group'>
          <BaseTheme
            id='setting-theme-shelob'
            theme='shelob'
            value='shelob'
            label='Shelob'
            checked={this.currentTheme === 'shelob'}
            onInput={(e: Event) => this.onInput(e)}
          />
          <BaseTheme
            id='setting-theme-denethor'
            theme='denethor'
            value='denethor'
            label='Denethor'
            checked={this.currentTheme === 'denethor'}
            onInput={(e: Event) => this.onInput(e)}
          />
          <BaseTheme
            id='setting-theme-quickbeam'
            theme='quickbeam'
            value='quickbeam'
            label='Quickbeam'
            checked={this.currentTheme === 'quickbeam'}
            onInput={(e: Event) => this.onInput(e)}
          />
          <BaseTheme
            id='setting-theme-shadowfax'
            theme='shadowfax'
            value='shadowfax'
            label='Shadowfax'
            checked={this.currentTheme === 'shadowfax'}
            onInput={(e: Event) => this.onInput(e)}
          />
          <BaseTheme
            id='setting-theme-grima'
            theme='grima'
            value='grima'
            label='Grima'
            checked={this.currentTheme === 'grima'}
            onInput={(e: Event) => this.onInput(e)}
          />
        </div>
      </fieldset>
    );
  }
}
