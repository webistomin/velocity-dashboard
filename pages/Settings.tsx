import { VueComponent } from 'types/vue-components';
import { Component } from 'nuxt-property-decorator';
import { VNode } from 'vue';
import Settings from 'components/ui/Settings';

@Component({
  name: 'SettingsPage',
})
export default class SettingsPage extends VueComponent {
  render(): VNode {
    return (
      <main class='page-content page-content_offset'>
        <Settings />
      </main>
    );
  }
}
