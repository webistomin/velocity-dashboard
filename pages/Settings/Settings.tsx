import { VueComponent } from 'types/vue-components';
import { Component } from 'nuxt-property-decorator';
import { VNode } from 'vue';
import Settings from 'components/ui/Settings';

@Component({
  name: 'SettingsPage',
  middleware: ['auth-guard'],
})
export default class SettingsPage extends VueComponent {
  public head() {
    return {
      title: 'Velocity :: Settings',
    };
  }

  render(): VNode {
    return (
      <main class='page-content page-content_offset'>
        <Settings />
        <notifications group='settings' position='top left' animation-name='v-popup-fade-left' />
      </main>
    );
  }
}
