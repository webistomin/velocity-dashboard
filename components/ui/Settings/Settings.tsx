import { VueComponent } from 'types/vue-components';
import { Component, Watch } from 'nuxt-property-decorator';
import Vue, { VNode } from 'vue';
import { State } from 'vuex-class';

import BaseTitle from 'components/base/BaseTitle';
import BaseBlock from 'components/base/BaseBlock';
import { IProfileUpdateResponseBody } from 'common/types/profile/update';
import { IUserInterface, IUserInterfaceDB } from 'common/types/user/user-schema';
import { SiteThemes } from 'common/types/theme/site-themes';
import BaseButton from 'components/base/BaseButton';
import { serverUrls } from 'common/urls/serverUrls';
import SettingsData from './SettingsData';
import SettingsNotifications from './SettingsNotifications';
import SettingsTheme from './SettingsTheme';

import './Settings.sass';

export interface IUserSettings {
  info: {
    email: IUserInterface['email'];
    firstName: IUserInterface['firstName'];
    lastName: IUserInterface['lastName'];
    dob: IUserInterface['dob'];
    currentPassword: IUserInterface['password'];
    newPassword: IUserInterface['password'];
    confirmPassword: IUserInterface['password'];
    bio: IUserInterface['bio'];
    location: IUserInterface['location'];
  };
  notifications: {
    isEmailNotificationsEnabled: IUserInterface['notifications']['isEmailNotificationsEnabled'];
    isPushNotificationsEnabled: IUserInterface['notifications']['isPushNotificationsEnabled'];
    isMonthlyNotificationsEnabled: IUserInterface['notifications']['isMonthlyNotificationsEnabled'];
    isQuarterNotificationsEnabled: IUserInterface['notifications']['isQuarterNotificationsEnabled'];
  };
  theme: IUserInterface['theme'];
}

@Component({
  name: 'Settings',
})
export default class Settings extends VueComponent {
  public $refs!: Vue['$refs'] & {
    settingsDataForm: SettingsData;
  };

  public isLoading: boolean = false;

  public settingsForm: IUserSettings = {
    info: {
      email: '',
      firstName: '',
      lastName: '',
      dob: '',
      currentPassword: '',
      newPassword: '',
      confirmPassword: '',
      bio: '',
      location: '',
    },
    notifications: {
      isEmailNotificationsEnabled: false,
      isPushNotificationsEnabled: false,
      isMonthlyNotificationsEnabled: false,
      isQuarterNotificationsEnabled: false,
    },
    theme: SiteThemes.SHELOB,
  };

  @State((state) => state.auth.user)
  getAuthUser!: IUserInterfaceDB;

  created(): void {
    const authUser = this.getAuthUser;

    const userInfo = this.settingsForm.info;
    const userNotification = this.settingsForm.notifications;

    userInfo.email = authUser.email;
    userInfo.firstName = authUser.firstName;
    userInfo.lastName = authUser.lastName;
    userInfo.dob = authUser.dob;
    userInfo.bio = authUser.bio;

    const notificationsKeys = Object.keys(userNotification) as Array<keyof IUserInterface['notifications']>;
    notificationsKeys.forEach((field) => {
      userNotification[field] = authUser.notifications[field];
    });

    this.settingsForm.theme = authUser.theme;
  }

  public onChangeSettings(info: IUserSettings['info']): void {
    this.settingsForm.info = info;
  }

  public onSelectNotification(notifications: IUserSettings['notifications']): void {
    this.settingsForm.notifications = notifications;
  }

  public onThemeSelect(theme: SiteThemes): void {
    this.settingsForm.theme = theme;
    const root = document.documentElement;
    root.className = `theme theme_${theme}`;
  }

  public async onSubmit(event: Event) {
    event.preventDefault();

    const validateState = this.$refs.settingsDataForm.touch();
    if (validateState.$anyError) {
      /**
       * Scroll to first error
       */
      await this.$nextTick();
      const firstError = document.querySelector('[id^=errors-settings]');
      if (firstError) {
        firstError.scrollIntoView({
          behavior: 'smooth',
          block: 'end',
        });
      }
    } else {
      try {
        this.isLoading = true;
        const data = this.settingsForm;
        const response: IProfileUpdateResponseBody = await this.$axios.$post(serverUrls.profile.update, data);

        if (response.success) {
          localStorage.setItem('theme', data.theme);

          if (response.token) {
            await this.$auth.setUserToken(response.token);
          } else {
            await this.$auth.fetchUser();
          }

          this.$notify({
            group: 'settings',
            type: 'success',
            title: 'Update successful',
            text: response.message,
            duration: 3000,
          });
        }

        this.isLoading = false;
      } catch (error) {
        this.isLoading = false;
        const response: IProfileUpdateResponseBody = error.response.data;
        this.$notify({
          group: 'settings',
          type: 'error',
          title: 'Error during update',
          text: response.message,
          duration: 3000,
        });
      }
    }
  }

  public beforeDestroy(): void {
    const theme = this.getAuthUser.theme;
    this.onThemeSelect(theme);
  }

  public render(): VNode {
    return (
      <section class='settings'>
        <div class='container settings__container'>
          <BaseTitle class='settings__title' level={3}>
            Settings
          </BaseTitle>
          <form class='settings__grid' onSubmit={this.onSubmit}>
            <BaseBlock title='Personal Data'>
              <SettingsData
                onChangeSettings={this.onChangeSettings}
                info={this.settingsForm.info}
                ref='settingsDataForm'
              />
            </BaseBlock>
            <BaseBlock title='Notifications'>
              <SettingsNotifications
                onSelectNotification={this.onSelectNotification}
                notifications={this.settingsForm.notifications}
              />
            </BaseBlock>
            <BaseBlock title='Theme'>
              <SettingsTheme onThemeSelect={this.onThemeSelect} theme={this.settingsForm.theme} />
            </BaseBlock>
            <BaseButton class='settings__submit' type='submit' isLoading={this.isLoading}>
              Save changes
            </BaseButton>
            <pre>{JSON.stringify(this.$v)}</pre>
          </form>
        </div>
      </section>
    );
  }
}
