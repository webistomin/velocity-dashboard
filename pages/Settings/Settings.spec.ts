import { shallowMount } from '@vue/test-utils';
import Settings from './Settings';

describe('Settings page', () => {
  const factory = (propsData = {}, options = {}) => {
    return shallowMount(Settings, {
      propsData: {
        ...propsData,
      },
      stubs: ['notifications'],
      ...options,
    });
  };

  it('Matches snapshot', () => {
    const wrapper = factory().html();
    expect(wrapper).toMatchSnapshot();
  });
});
