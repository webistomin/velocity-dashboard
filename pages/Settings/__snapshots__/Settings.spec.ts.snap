// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings page Matches snapshot 1`] = `
"<main class=\\"page-content page-content_offset\\">
  <settings-stub></settings-stub>
  <notifications-stub group=\\"settings\\" position=\\"top left\\" animation-name=\\"v-popup-fade-left\\"></notifications-stub>
</main>"
`;
